### PR TITLE
Cgs

### PR DESCRIPTION
**Pull request by robbje**
This patch refactors function listing/printing code.

I believe it's more maintainable, easier to read, removes dead code and is
less prone to bugs, while conserving existing functionality.

### CODE REVIEW

**Comment by crowell** [comment]
please use tabs instead of spaces

File: libr/core/anal.c [hunk 1]
```diff
@@ -1375,7 +1375,7 @@ R_API int r_core_anal_fcn_list_size(RCore *core) {
 	RAnalFunction *fcn;
 	RListIter *iter;
 	ut32 total = 0;
-
+
```
Comment (crowell) [inline libr/core/anal.c]: kill tab here

File: libr/core/anal.c [hunk 2]
```diff
@@ -1388,296 +1388,410 @@ static int cmpfcn(const void *_a, const void *_b) {
...
+		fcn->meta.min = min;
+		fcn->meta.max = max;
+	}
+	// TODO: Determine sgnc, sgec
+	return 0;
+}
+
+static char *r_core_get_function_name(RCore *core, RAnalFunction *fcn) {
+    bool demangle;
+    const char *lang;
```
Comment (crowell) [inline libr/core/anal.c]: tab

File: libr/core/anal.c [hunk 3]
```diff
@@ -1388,296 +1388,410 @@ static int cmpfcn(const void *_a, const void *_b) {
...
+		}
+		fcn->meta.min = min;
+		fcn->meta.max = max;
+	}
+	// TODO: Determine sgnc, sgec
+	return 0;
+}
+
+static char *r_core_get_function_name(RCore *core, RAnalFunction *fcn) {
+    bool demangle;
```
Comment (crowell) [inline libr/core/anal.c]: tab

File: libr/core/anal.c [hunk 4]
```diff
@@ -1388,296 +1388,410 @@ static int cmpfcn(const void *_a, const void *_b) {
...
+    char *name = strdup (fcn->name ? fcn->name : "");
+    if (demangle) {
+        char *tmp;
+        tmp = r_bin_demangle (core->bin->cur, lang, name);
+        if (tmp) {
+            free(name);
+            name = tmp;
+        }
+    } 
+    return name;
```
Comment (crowell) [inline libr/core/anal.c]: potential UAF here

File: libr/core/cmd_anal.c [hunk 5]
```diff
@@ -972,33 +972,39 @@ static int cmd_anal_fcn(RCore *core, const char *input) {
...
-		case 'j': r_core_anal_fcn_list (core, input + 3, 'j'); break; // "afij"
-		case '*': r_core_anal_fcn_list (core, input + 3, 1); break;   // "afi*"
-		default: r_core_anal_fcn_list (core, input + 2, 0); break;
+		case '?':
+        	eprintf ("Usage: afi[jl*] <addr>\n"); break;
+        	eprintf ("afij - function info in json format\n");
+			eprintf ("afil - verbose function info\n");
+        	eprintf ("afi* - function, variables and arguments\n");
+		case 'j':   // "afij"
+        case 'l':   // "afil"
```
Comment (crowell) [inline libr/core/cmd_anal.c]: tabs

File: libr/core/cmd_anal.c [hunk 6]
```diff
@@ -972,33 +972,39 @@ static int cmd_anal_fcn(RCore *core, const char *input) {
...
+        	eprintf ("aflj - list functions in json format\n");
+			eprintf ("afll - list functions in verbose mode\n");
+        	eprintf ("aflq - list functions in 'quiet' mode\n");
+        	eprintf ("afls - print sum of sizes of all functions\n");
 			break;
-		case 'a':
-		case '*':
-		case 'j':
-		case 'q': // "aflq"
+        case 'j':
```
Comment (crowell) [inline libr/core/cmd_anal.c]: tabs

**Comment by radare** [comment]
do not merge in branches

File: libr/core/anal.c [hunk 7]
```diff
@@ -1373,294 +1373,410 @@ static int cmpfcn(const void *_a, const void *_b) {
...
 	lang = demangle ? r_config_get (core->config, "bin.lang") : NULL;
-	bool use_color = r_config_get_i (core->config, "scr.color");
 
-	if (input && *input)
-		addr = r_num_math (core->num, input + 1);
-	else
-		addr = core->offset;
+    char *name = strdup (fcn->name ? fcn->name : "");
+    if (demangle) {
+        char *tmp;
```
Comment (radare) [inline libr/core/anal.c]: assign in the same line as the definition

File: libr/core/anal.c [hunk 8]
```diff
@@ -1373,294 +1373,410 @@ static int cmpfcn(const void *_a, const void *_b) {
...
+        tmp = r_bin_demangle (core->bin->cur, lang, name);
+        if (tmp) {
+            free(name);
```
Comment (radare) [inline libr/core/anal.c]: missing space

File: libr/core/anal.c [hunk 9]
```diff
@@ -1373,294 +1373,410 @@ static int cmpfcn(const void *_a, const void *_b) {
...
+
+    return 0;
+}
+
+R_API int r_core_anal_fcn_print_default(RCore *core, RAnalFunction *fcn, bool quiet) {
+    if (quiet) {
+        r_cons_printf ("0x%08"PFMT64x" ", fcn->addr);
+    } else {
+        char *name = r_core_get_function_name (core, fcn);
+        char *msg;
```
Comment (radare) [inline libr/core/anal.c]: define in the same line as `char *msg, *name = ...`

File: libr/core/anal.c [hunk 10]
```diff
@@ -1373,294 +1373,410 @@ static int cmpfcn(const void *_a, const void *_b) {
...
+    
+        int realsize = r_anal_fcn_realsize (fcn);
+        int size = r_anal_fcn_size (fcn);
+        if(realsize == size) {
```
Comment (radare) [inline libr/core/anal.c]: missing space before (

File: libr/core/anal.c [hunk 11]
```diff
@@ -1373,294 +1373,410 @@ static int cmpfcn(const void *_a, const void *_b) {
...
+            msg = r_str_newf ("%-12d", size);
+        } else {
+            msg = r_str_newf ("%-4d -> %-4d", size, realsize);
+        }
+
```
Comment (radare) [inline libr/core/anal.c]: unnecessary empty lines

File: libr/core/anal.c [hunk 12]
```diff
@@ -1373,294 +1373,410 @@ static int cmpfcn(const void *_a, const void *_b) {
...
+        if(realsize == size) {
+            msg = r_str_newf ("%-12d", size);
+        } else {
+            msg = r_str_newf ("%-4d -> %-4d", size, realsize);
+        }
+    
+        r_cons_printf ("0x%08"PFMT64x" %4d %4s %s\n",
+        		fcn->addr, r_list_length (fcn->bbs), msg, name);
+    
+        R_FREE (name);
```
Comment (radare) [inline libr/core/anal.c]: just use free(), values are not used anymore after this

File: libr/core/anal.c [hunk 13]
```diff
@@ -1373,294 +1373,410 @@ static int cmpfcn(const void *_a, const void *_b) {
...
+	r_cons_printf (",\"calltype\":\"%s\"", r_anal_cc_type2str (fcn->call));
+	r_cons_printf (",\"type\":\"%s\"",
+			fcn->type==R_ANAL_FCN_TYPE_SYM?"sym":
+			fcn->type==R_ANAL_FCN_TYPE_IMP?"imp":"fcn");
+	if (fcn->type == R_ANAL_FCN_TYPE_FCN || fcn->type == R_ANAL_FCN_TYPE_SYM)
+		r_cons_printf (",\"diff\":\"%s\"",
+				fcn->diff->type==R_ANAL_DIFF_TYPE_MATCH?"MATCH":
+				fcn->diff->type==R_ANAL_DIFF_TYPE_UNMATCH?"UNMATCH":"NEW");
+
+    int first = 1;
```
Comment (radare) [inline libr/core/anal.c]: define at the begining of function

File: libr/core/anal.c [hunk 14]
```diff
@@ -1373,294 +1373,410 @@ static int cmpfcn(const void *_a, const void *_b) {
...
+			refi->type == R_ANAL_REF_TYPE_CALL)
+			r_cons_printf ("0x%08"PFMT64x" %c ", refi->addr,
+					refi->type==R_ANAL_REF_TYPE_CALL?'C':'J');
+
+	r_cons_printf ("\n data-xrefs: ");
+	r_list_foreach (fcn->xrefs, iter, refi)
+		if (refi->type == R_ANAL_REF_TYPE_DATA)
+			r_cons_printf ("0x%08"PFMT64x" ", refi->addr);
+
+	if (fcn->type==R_ANAL_FCN_TYPE_FCN || fcn->type==R_ANAL_FCN_TYPE_SYM) {
```
Comment (radare) [inline libr/core/anal.c]: missing spaces around ==

File: libr/core/anal.c [hunk 15]
```diff
@@ -1373,294 +1373,410 @@ static int cmpfcn(const void *_a, const void *_b) {
...
+		r_cons_printf ("\n diff: type: %s",
+				fcn->diff->type==R_ANAL_DIFF_TYPE_MATCH?"match":
+				fcn->diff->type==R_ANAL_DIFF_TYPE_UNMATCH?"unmatch":"new");
+		if (fcn->diff->addr != -1)
+			r_cons_printf (" addr: 0x%"PFMT64x, fcn->diff->addr);
+		if (fcn->diff->name != NULL)
+			r_cons_printf (" function: %s",
+				fcn->diff->name);
+	}
+    R_FREE (name)
```
Comment (radare) [inline libr/core/anal.c]: use free

File: libr/core/anal.c [hunk 16]
```diff
@@ -1373,294 +1373,410 @@ static int cmpfcn(const void *_a, const void *_b) {
...
+        case 1:
+            r_core_anal_fcn_list_legacy(core, fcns);
+            break;
+        default:
+            r_core_anal_fcn_list_default (core, fcns, false);
+            break;
+    }
+    if (has_filter) {
+        // The list does not own the its members, so don't purge.
+        R_FREE(fcns);
```
Comment (radare) [inline libr/core/anal.c]: use free

File: libr/core/anal.c [hunk 17]
```diff
@@ -1743,12 +1743,11 @@ R_API int r_core_anal_fcn_list(RCore *core, const char *input, int rad) {
 	r_list_sort (fcns, &cmpfcn);
 	r_core_anal_fcn_gather_metadata(fcns);
 	
-	bool has_filter = (input && *input);
+	ut64 addr;
+	addr = core->offset;
+	bool has_filter = input;
```
Comment (radare) [inline libr/core/anal.c]: wtf?

File: libr/core/anal.c [hunk 18]
```diff
@@ -1743,12 +1743,11 @@ R_API int r_core_anal_fcn_list(RCore *core, const char *input, int rad) {
 	r_list_sort (fcns, &cmpfcn);
 	r_core_anal_fcn_gather_metadata(fcns);
```
Comment (radare) [inline libr/core/anal.c]: missing space before (

File: libr/core/anal.c [hunk 19]
```diff
@@ -1758,7 +1757,7 @@ R_API int r_core_anal_fcn_list(RCore *core, const char *input, int rad) {
 		RListIter *iter;
 		RAnalFunction *fcn;
 		r_list_foreach(core->anal->fcns, iter, fcn) {
-			if (in_function(fcn, addr) || (!strcmp(input, fcn->name))) {
+			if (in_function(fcn, addr) || (!strcmp(input + 1, fcn->name))) {
```
Comment (radare) [inline libr/core/anal.c]: missing spaces before (

File: libr/core/anal.c [hunk 20]
```diff
@@ -1758,7 +1757,7 @@ R_API int r_core_anal_fcn_list(RCore *core, const char *input, int rad) {
...
 				r_list_append(fcns, fcn);
```
Comment (radare) [inline libr/core/anal.c]: same here

**Comment by ret2libc** [comment]
@radare @robbje I think since you changed some function to be `static`, they don't need anymore the `r_core_....` prefix (in general, if I'm not wrong, `r_` prefix is for R_API functions). We can make those names shorter.

**Comment by radare** [comment]
agree on that @ret2libc suggestion. mind to shorten those function names? i'll merge after that. thanks!

**Comment by robbje** [comment]
I'll do that later tonight. I appreciate all the helpful comments. Thanks!

**Comment by radare** [comment]
Thanks!

**Pull request merged**